sdk/iot/pnp/inc/az_iot_pnp_client.h

```diff
@@ -0,0 +1,334 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+
+/**
+ * @file az_iot_pnp_client.h
+ *
+ * @brief definition for the Azure IoT PnP device SDK.
+ */
+
+#ifndef _az_IOT_PNP_CLIENT_H
+#define _az_IOT_PNP_CLIENT_H
+
+#include <az_result.h>
+#include <az_span.h>
+
+// NOTE: az_iot_hub_client.h is included for internal use only.
+#include <az_iot_hub_client.h>
+
+#include <stdbool.h>
+#include <stdint.h>
+
+#include <_az_cfg_prefix.h>
+
+/**
+ * @brief Azure IoT PnP Client options.
+ *
+ */
+typedef struct az_iot_pnp_client_options
+{
+  az_span user_agent; /**< The user-agent is a formatted string that will be used for Azure IoT
+                         usage statistics. */
+} az_iot_pnp_client_options;
+
+/**
+ * @brief Azure IoT PnP Client.
+ *
+ */
+typedef struct az_iot_pnp_client
+{
+  struct
+  {
+    az_iot_hub_client iot_hub_client;
+    az_span root_interface_name;
+  } _internal;
+} az_iot_pnp_client;
+
+/**
+ * @brief Gets the default Azure IoT PnP Client options.
+ * @details Call this to obtain an initialized #az_iot_pnp_client_options structure that can be
+ *          afterwards modified and passed to #az_iot_pnp_client_init.
+ *
+ * @return #az_iot_pnp_client_options.
+ */
+AZ_NODISCARD az_iot_pnp_client_options az_iot_pnp_client_options_default();
+
+/**
+ * @brief Initializes an Azure IoT PnP Client.
+ *
+ * @param[out] client The #az_iot_pnp_client to use for this call.
+ * @param[in] iot_hub_hostname The IoT Hub Hostname.
+ * @param[in] device_id The Device ID.
+ * @param[in] root_interface_name The root interface of the #az_iot_pnp_client.
+ * @param[in] options A reference to an #az_iot_pnp_client_options structure. Can be NULL.
+ * @return #az_result
+ */
+AZ_NODISCARD az_result az_iot_pnp_client_init(
+    az_iot_pnp_client* client,
+    az_span iot_hub_hostname,
+    az_span device_id,
+    az_span root_interface_name,
```

Comment:
Maybe these spans should be copied into byte buffers that are in the client to make them immutable and the customer doesn't have to keep the spans around. Although it's not a problem if the span are over literal strings.

Reply:
Fixed, never dynamic.  

The root_interface defines what the device is in the 1st place, e.g. "Constoso-Air-Conditioner" and reflects modeled DTDL.  The only time it's used *today* is on the initial MQTT CONNECT packet in building user name.  However, this could potentially be different in a future binding so I don't want to have this passed in the `get_username` stuff.  And it's fundamental to what the device is it seems to make sense in the pnp_client itself.

Reply:
@JeffreyRichter - the vast majority of the time these will be string literals, since it'll be hard-coded into the application itself.  It knows it's always a "Contoso-....-whatever", this being dynamically built would be surprising.  So I think the simpler mechanism in there today is OK.

Reply:
ok

Reply:
> The root_interface defines what the device is in the 1st place, e.g. "Constoso-Air-Conditioner" and reflects modeled DTDL.

It really makes no sense to me to have it here - the Contoso-Air-Conditioner cannot suddenly become the Contoso-High-Performance-Computer overnight. The association between a device and it's type should be done one time only: at provisioning.
Allowing this to change from the device side, dynamically, presents all sorts of issues.